File: gemsfx/src/main/java/com/dlsc/gemsfx/treeview/TreeNodeView.java
package com.dlsc.gemsfx.treeview;

import com.dlsc.gemsfx.treeview.link.LinkStrategy;
import com.dlsc.gemsfx.treeview.link.StraightLineLink;
import javafx.beans.property.DoubleProperty;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.css.CssMetaData;
import javafx.css.PseudoClass;
import javafx.css.Styleable;
import javafx.css.StyleableDoubleProperty;
import javafx.css.StyleableObjectProperty;
import javafx.css.StyleableProperty;
import javafx.css.converter.EnumConverter;
import javafx.css.converter.SizeConverter;
import javafx.geometry.HPos;
import javafx.geometry.VPos;
import javafx.scene.Node;
import javafx.scene.control.Control;
import javafx.scene.control.Label;
import javafx.scene.control.Skin;
import javafx.util.Callback;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;


/**
 * A visual control for displaying trees.
 * <p>
 * Built on the {@link TreeNode} class, this control visualizes hierarchical structures, allowing nodes to have children.
 * <p>
 * Customizable in layout, alignment, and style, it's ideal for representing data like file systems or organizational charts.
 */
public class TreeNodeView<T> extends Control {
    private static final int DEFAULT_CELL_WIDTH = 60;
    private static final int DEFAULT_CELL_HEIGHT = 30;
    private static final double DEFAULT_VGAP = 50;
    private static final double DEFAULT_HGAP = 20;
    private static final double DEFAULT_NODE_LINE_GAP = 10;
    private static final VPos DEFAULT_ROW_ALIGNMENT = VPos.CENTER;
    private static final HPos DEFAULT_COLUMN_ALIGNMENT = HPos.CENTER;
    private static final LayoutType DEFAULT_LAYOUT_TYPE = LayoutType.REGULAR;
    private static final LayoutDirection DEFAULT_LAYOUT_DIRECTION = LayoutDirection.TOP_TO_BOTTOM;
    private static final PseudoClass LTR_PSEUDOCLASS_STATE = PseudoClass.getPseudoClass("ltr");
    private static final PseudoClass RTL_PSEUDOCLASS_STATE = PseudoClass.getPseudoClass("rtl");
    private static final PseudoClass TTB_PSEUDOCLASS_STATE = PseudoClass.getPseudoClass("ttb");
    private static final PseudoClass BTT_PSEUDOCLASS_STATE = PseudoClass.getPseudoClass("btt");
    private static final String DEFAULT_STYLE_CLASS = "tree-node-view";

    public TreeNodeView() {
        getStyleClass().add(DEFAULT_STYLE_CLASS);
        layoutDirection.addListener(it -> activateDirectionPseudoClass());
    }

    public TreeNodeView(TreeNode<T> root) {
        this();
        setRoot(root);
    }

    private void activateDirectionPseudoClass() {
        pseudoClassStateChanged(LTR_PSEUDOCLASS_STATE, getLayoutDirection() == LayoutDirection.LEFT_TO_RIGHT);
        pseudoClassStateChanged(RTL_PSEUDOCLASS_STATE, getLayoutDirection() == LayoutDirection.RIGHT_TO_LEFT);
        pseudoClassStateChanged(TTB_PSEUDOCLASS_STATE, getLayoutDirection() == LayoutDirection.TOP_TO_BOTTOM);
        pseudoClassStateChanged(BTT_PSEUDOCLASS_STATE, getLayoutDirection() == LayoutDirection.BOTTOM_TO_TOP);
    }

    @Override
    protected Skin<?> createDefaultSkin() {
        return new TreeNodeViewSkin<>(this);
    }

    @Override
    public String getUserAgentStylesheet() {
        return Objects.requireNonNull(getClass().getResource("tree-view.css")).toExternalForm();
    }

    private final ObjectProperty<Callback<T, TreeNodeCell<T>>> cellFactory = new SimpleObjectProperty<>(this, "cellFactory", TreeNodeCell::new);

    public final void setCellFactory(Callback<T, TreeNodeCell<T>> value) {
        cellFactoryProperty().set(value);
    }

    public final Callback<T, TreeNodeCell<T>> getCellFactory() {
        return cellFactory.get();
    }

    public final ObjectProperty<Callback<T, TreeNodeCell<T>>> cellFactoryProperty() {
        return cellFactory;
    }

    private final ObjectProperty<TreeNode<T>> root = new SimpleObjectProperty<>(this, "root");

    public TreeNode<T> getRoot() {
        return root.get();
    }

    public ObjectProperty<TreeNode<T>> rootProperty() {
        return root;
    }

    public void setRoot(TreeNode<T> root) {
        this.root.set(root);
    }

    /**
     * Alignment on the same level;
     * When the layout direction is btt or ttb,
     * this attribute indicates the alignment of nodes at the same level;
     */
    private final StyleableObjectProperty<VPos> rowAlignment = new StyleableObjectProperty<>(DEFAULT_ROW_ALIGNMENT) {
        @Override
        public CssMetaData<? extends Styleable, VPos> getCssMetaData() {
            return StyleableProperties.ROW_ALIGNMENT;
        }

        @Override
        public Object getBean() {
            return TreeNodeView.this;
        }

        @Override
        public String getName() {
            return "rowAlignment";
        }
    };

    public VPos getRowAlignment() {
        return rowAlignment.get();
    }

    public ObjectProperty<VPos> rowAlignmentProperty() {
        return rowAlignment;
    }

    public void setRowAlignment(VPos rowAlignment) {
        this.rowAlignment.set(rowAlignment);
    }

    /**
     * Alignment on the same level;
     * When the layout direction is ltr or rtl,
     * this attribute indicates the alignment of nodes at the same level;
     */
    private final StyleableObjectProperty<HPos> columnAlignment = new StyleableObjectProperty<>(DEFAULT_COLUMN_ALIGNMENT) {
        @Override
        public CssMetaData<? extends Styleable, HPos> getCssMetaData() {
            return StyleableProperties.COLUMN_ALIGNMENT;
        }

        @Override
        public Object getBean() {
            return TreeNodeView.this;
        }

        @Override
        public String getName() {
            return "columnAlignment";
        }
    };

    public HPos getColumnAlignment() {
        return columnAlignment.get();
    }

    public ObjectProperty<HPos> columnAlignmentProperty() {
        return columnAlignment;
    }

    public void setColumnAlignment(HPos columnAlignment) {
        this.columnAlignment.set(columnAlignment);
    }

    private final DoubleProperty cellWidth = new StyleableDoubleProperty(DEFAULT_CELL_WIDTH) {
        @Override
        public CssMetaData<? extends Styleable, Number> getCssMetaData() {
            return StyleableProperties.CELL_WIDTH;
        }

        @Override
        public Object getBean() {
            return TreeNodeView.this;
        }

        @Override
        public String getName() {
            return "cellWidth";
        }
    };

    public double getCellWidth() {
        return cellWidth.get();
    }

    public DoubleProperty cellWidthProperty() {
        return cellWidth;
    }

    public void setCellWidth(double cellWidth) {
        this.cellWidth.set(cellWidth);
    }

    private final DoubleProperty cellHeight = new StyleableDoubleProperty(DEFAULT_CELL_HEIGHT) {
        @Override
        public CssMetaData<? extends Styleable, Number> getCssMetaData() {
            return StyleableProperties.CELL_HEIGHT;
        }

        @Override
        public Object getBean() {
            return TreeNodeView.this;
        }

        @Override
        public String getName() {
            return "cellHeight";
        }

    };

    public double getCellHeight() {
        return cellHeight.get();
    }

    public DoubleProperty cellHeightProperty() {
        return cellHeight;
    }

    public void setCellHeight(double cellHeight) {
        this.cellHeight.set(cellHeight);
    }

    private final DoubleProperty hgap = new StyleableDoubleProperty(DEFAULT_HGAP) {
        @Override
        public CssMetaData<? extends Styleable, Number> getCssMetaData() {
            return StyleableProperties.H_GAP;
        }

        @Override
        public Object getBean() {
            return TreeNodeView.this;
        }

        @Override
        public String getName() {
            return "hgap";
        }
    };

    public double getHgap() {
        return hgap.get();
    }

    public DoubleProperty hgapProperty() {
        return hgap;
    }

    public void setHgap(double hgap) {
        this.hgap.set(hgap);
    }

    private final DoubleProperty vgap = new StyleableDoubleProperty(DEFAULT_VGAP) {
        @Override
        public CssMetaData<? extends Styleable, Number> getCssMetaData() {
            return StyleableProperties.V_GAP;
        }

        @Override
        public Object getBean() {
            return TreeNodeView.this;
        }

        @Override
        public String getName() {
            return "vgap";
        }
    };

    public double getVgap() {
        return vgap.get();
    }

    public DoubleProperty vgapProperty() {
        return vgap;
    }

    public void setVgap(double vgap) {
        this.vgap.set(vgap);
    }

    private final DoubleProperty nodeLineGap = new StyleableDoubleProperty(DEFAULT_NODE_LINE_GAP) {
        @Override
        public CssMetaData<? extends Styleable, Number> getCssMetaData() {
            return StyleableProperties.NODE_LINE_GAP;
        }

        @Override
        public Object getBean() {
            return TreeNodeView.this;
        }

        @Override
        public String getName() {
            return "nodeLineGap";
        }
    };

    public double getNodeLineGap() {
        return nodeLineGap.get();
    }

    public DoubleProperty nodeLineGapProperty() {
        return nodeLineGap;
    }

    public void setNodeLineGap(double nodeLineGap) {
        this.nodeLineGap.set(nodeLineGap);
    }

    private final ObjectProperty<Node> placeholder = new SimpleObjectProperty<>(this, "placeholder", createDefaultPlaceholder());

    public Node getPlaceholder() {
        return placeholder.get();
    }

    public ObjectProperty<Node> placeholderProperty() {
        return placeholder;
    }

    public void setPlaceholder(Node placeholder) {
        this.placeholder.set(placeholder);
    }

    private final ObjectProperty<LinkStrategy<T>> linkStrategy = new SimpleObjectProperty<>(this, "linkStrategy", new StraightLineLink<>());

    public LinkStrategy<T> getLinkStrategy() {
        return linkStrategy.get();
    }

    public ObjectProperty<LinkStrategy<T>> linkStrategyProperty() {
        return linkStrategy;
    }

    public void setLinkStrategy(LinkStrategy<T> linkStrategy) {
        this.linkStrategy.set(linkStrategy);
    }

    private Node createDefaultPlaceholder() {
        Label label = new Label("No tree root.");
        label.getStyleClass().add("default-placeholder");
        return label;
    }

    public enum LayoutType {
        /**
         * The tree node view will be layout in a regular way. may be wider
         */
        REGULAR,
        /**
         * The tree node view will be layout in a compact way. may be narrower.
         */
        COMPACT;
    }

    private final ObjectProperty<LayoutType> layoutType = new StyleableObjectProperty<>(DEFAULT_LAYOUT_TYPE) {
        @Override
        public CssMetaData<? extends Styleable, LayoutType> getCssMetaData() {
            return StyleableProperties.LAYOUT_TYPE;
        }

        @Override
        public Object getBean() {
            return TreeNodeView.this;
        }

        @Override
        public String getName() {
            return "layoutType";
        }
    };

    public LayoutType getLayoutType() {
        return layoutType.get();
    }

    public ObjectProperty<LayoutType> layoutTypeProperty() {
        return layoutType;
    }

    public void setLayoutType(LayoutType layoutType) {
        this.layoutType.set(layoutType);
    }

    public enum LayoutDirection {
        /**
         * The tree node view will be layout from left to right.
         */
        LEFT_TO_RIGHT,
        /**
         * The tree node view will be layout from right to left.
         */
        RIGHT_TO_LEFT,
        /**
         * The tree node view will be layout from top to bottom.
         */
        TOP_TO_BOTTOM,
        /**
         * The tree node view will be layout from bottom to top.
         */
        BOTTOM_TO_TOP;
    }

    private final ObjectProperty<LayoutDirection> layoutDirection = new StyleableObjectProperty<>(DEFAULT_LAYOUT_DIRECTION) {
        @Override
        public CssMetaData<? extends Styleable, LayoutDirection> getCssMetaData() {
            return StyleableProperties.LAYOUT_DIRECTION;
        }

        @Override
        public Object getBean() {
            return TreeNodeView.this;
        }

        @Override
        public String getName() {
            return "layoutDirection";
        }
    };

    public LayoutDirection getLayoutDirection() {
        return layoutDirection.get();
    }

    public ObjectProperty<LayoutDirection> layoutDirectionProperty() {
        return layoutDirection;
    }

    public void setLayoutDirection(LayoutDirection layoutDirection) {
        this.layoutDirection.set(layoutDirection);
    }

    public void refresh() {
        Skin<?> skin = getSkin();
        if (skin instanceof TreeNodeViewSkin) {
            @SuppressWarnings("unchecked")
            TreeNodeViewSkin<T> treeNodeSkin = (TreeNodeViewSkin<T>) skin;
            treeNodeSkin.refresh();
        }
    }

    private static class StyleableProperties {

        public static final CssMetaData<TreeNodeView<?>, HPos> COLUMN_ALIGNMENT =
                new CssMetaData<>("-fx-column-alignment", new EnumConverter<>(HPos.class), DEFAULT_COLUMN_ALIGNMENT) {
                    @Override
                    public boolean isSettable(TreeNodeView<?> n) {
                        return !n.columnAlignment.isBound();
                    }

                    @Override
                    @SuppressWarnings("unchecked")
                    public StyleableProperty<HPos> getStyleableProperty(TreeNodeView<?> n) {
                        return (StyleableProperty<HPos>) n.columnAlignmentProperty();
                    }
                };
        public static final CssMetaData<TreeNodeView<?>, LayoutType> LAYOUT_TYPE =
                new CssMetaData<>("-fx-layout-type", new EnumConverter<>(LayoutType.class), DEFAULT_LAYOUT_TYPE) {
                    @Override
                    public boolean isSettable(TreeNodeView<?> styleable) {
                        return !styleable.layoutType.isBound();
                    }

                    @Override
                    @SuppressWarnings("unchecked")
                    public StyleableProperty<LayoutType> getStyleableProperty(TreeNodeView<?> styleable) {
                        return (StyleableProperty<LayoutType>) styleable.layoutTypeProperty();
                    }
                };


        private static final CssMetaData<TreeNodeView<?>, VPos> ROW_ALIGNMENT =
                new CssMetaData<>("-fx-row-alignment", new EnumConverter<>(VPos.class), DEFAULT_ROW_ALIGNMENT) {
                    @Override
                    public boolean isSettable(TreeNodeView<?> n) {
                        return !n.rowAlignment.isBound();
                    }

                    @Override
                    @SuppressWarnings("unchecked")
                    public StyleableProperty<VPos> getStyleableProperty(TreeNodeView<?> n) {
                        return (StyleableProperty<VPos>) n.rowAlignmentProperty();
                    }
                };

        public static final CssMetaData<TreeNodeView<?>, Number> CELL_WIDTH =
                new CssMetaData<>("-fx-cell-width", SizeConverter.getInstance(), DEFAULT_CELL_WIDTH) {
                    @Override
                    public boolean isSettable(TreeNodeView<?> styleable) {
                        return !styleable.cellWidth.isBound();
                    }

                    @Override
                    @SuppressWarnings("unchecked")
                    public StyleableProperty<Number> getStyleableProperty(TreeNodeView<?> styleable) {
                        return (StyleableProperty<Number>) styleable.cellWidthProperty();
                    }
                };

        public static final CssMetaData<TreeNodeView<?>, Number> CELL_HEIGHT =
                new CssMetaData<>("-fx-cell-height", SizeConverter.getInstance(), DEFAULT_CELL_HEIGHT) {
                    @Override
                    public boolean isSettable(TreeNodeView<?> styleable) {
                        return !styleable.cellHeight.isBound();
                    }

                    @Override
                    @SuppressWarnings("unchecked")
                    public StyleableProperty<Number> getStyleableProperty(TreeNodeView<?> styleable) {
                        return (StyleableProperty<Number>) styleable.cellHeightProperty();
                    }
                };

        public static final CssMetaData<TreeNodeView<?>, Number> H_GAP =
                new CssMetaData<>("-fx-hgap", SizeConverter.getInstance(), DEFAULT_HGAP) {
                    @Override
                    public boolean isSettable(TreeNodeView<?> styleable) {
                        return !styleable.hgap.isBound();
                    }

                    @Override
                    @SuppressWarnings("unchecked")
                    public StyleableProperty<Number> getStyleableProperty(TreeNodeView<?> styleable) {
                        return (StyleableProperty<Number>) styleable.hgapProperty();
                    }
                };

        public static final CssMetaData<TreeNodeView<?>, Number> V_GAP =
                new CssMetaData<>("-fx-vgap", SizeConverter.getInstance(), DEFAULT_VGAP) {
                    @Override
                    public boolean isSettable(TreeNodeView<?> styleable) {
                        return !styleable.vgap.isBound();
                    }

                    @Override
                    @SuppressWarnings("unchecked")
                    public StyleableProperty<Number> getStyleableProperty(TreeNodeView<?> styleable) {
                        return (StyleableProperty<Number>) styleable.vgapProperty();
                    }
                };

        public static final CssMetaData<TreeNodeView<?>, Number> NODE_LINE_GAP =
                new CssMetaData<>("-fx-node-line-gap", SizeConverter.getInstance(), DEFAULT_NODE_LINE_GAP) {
                    @Override
                    public boolean isSettable(TreeNodeView<?> styleable) {
                        return !styleable.nodeLineGap.isBound();
                    }

                    @Override
                    @SuppressWarnings("unchecked")
                    public StyleableProperty<Number> getStyleableProperty(TreeNodeView<?> styleable) {
                        return (StyleableProperty<Number>) styleable.nodeLineGapProperty();
                    }
                };

        public static final CssMetaData<TreeNodeView<?>, LayoutDirection> LAYOUT_DIRECTION =
                new CssMetaData<>("-fx-layout-direction", new EnumConverter<>(LayoutDirection.class), DEFAULT_LAYOUT_DIRECTION) {
                    @Override
                    public boolean isSettable(TreeNodeView<?> styleable) {
                        return !styleable.layoutDirection.isBound();
                    }

                    @Override
                    @SuppressWarnings("unchecked")
                    public StyleableProperty<LayoutDirection> getStyleableProperty(TreeNodeView<?> styleable) {
                        return (StyleableProperty<LayoutDirection>) styleable.layoutDirectionProperty();
                    }
                };
        private static final List<CssMetaData<? extends Styleable, ?>> STYLEABLES;

        static {
            List<CssMetaData<? extends Styleable, ?>> styleables = new ArrayList<>(Control.getClassCssMetaData());
            Collections.addAll(styleables, ROW_ALIGNMENT, COLUMN_ALIGNMENT, CELL_WIDTH, CELL_HEIGHT, H_GAP, V_GAP, NODE_LINE_GAP, LAYOUT_TYPE, LAYOUT_DIRECTION);
            STYLEABLES = Collections.unmodifiableList(styleables);
        }
    }

    @Override
    protected List<CssMetaData<? extends Styleable, ?>> getControlCssMetaData() {
        return getClassCssMetaData();
    }

    public static List<CssMetaData<? extends Styleable, ?>> getClassCssMetaData() {
        return StyleableProperties.STYLEABLES;
    }
}

File: docs/manuals/src/manuals/tree_node_view.py
"""Content of the TreeNodeView developer manual."""

from manualkit import (
    Bullets, Callout, Chapter, Code, Figure, Manual, Numbered, Para,
    Property, PropertyTable, Section, Table,
)

G = "tree-node-view"

MANUAL = Manual(
    control='TreeNodeView',
    package='com.dlsc.gemsfx.treeview',
    subtitle='A layout control for visual tree diagrams',
    abstract='TreeNodeView renders a TreeNode model as a node-link diagram with configurable layout direction, regular or compact layout, cell factory, spacing, alignment and link strategies.',
    cover_svg=f"{G}/cover.svg",
    cover_caption='Generated cartoon overview of TreeNodeView.',
    chapters=[
        Chapter("Introduction", [
            Para("<b>TreeNodeView</b> lives in <font face='Courier'>com.dlsc.gemsfx.treeview</font>. It does not use JavaFX TreeItem; instead, its model is <font face='Courier'>TreeNode</font>, which stores children, optional linked nodes, expansion state and optional per-node dimensions."),
            Section("Key features"),
            Bullets([
                'Supports TOP_TO_BOTTOM, BOTTOM_TO_TOP, LEFT_TO_RIGHT and RIGHT_TO_LEFT layout directions.',
                'Supports REGULAR and COMPACT layout algorithms.',
                'Default cells are TreeNodeCell instances with label and disclosure arrow.',
                'Links are pluggable through LinkStrategy implementations.',
                'TreeNode.name creates deterministic style classes for node links.'
            ]),
            Section("Maven dependency"),
            Code("""<dependency>
    <groupId>com.dlsc.gemsfx</groupId>
    <artifactId>gemsfx</artifactId>
    <version>4.4.1</version>
</dependency>""", caption="Use package <font face='Courier'>com.dlsc.gemsfx.treeview</font>."),
        ]),
        Chapter("Getting started", [
            Para("The snippet below uses only APIs verified in the source and demo code."),
            Code('TreeNode<String> root = new TreeNode<>("Root");\nroot.getChildren().addAll(new TreeNode<>("A"), new TreeNode<>("B"));\n\nTreeNodeView<String> view = new TreeNodeView<>(root);\nview.setLayoutDirection(TreeNodeView.LayoutDirection.TOP_TO_BOTTOM);\nview.setLayoutType(TreeNodeView.LayoutType.REGULAR);\nview.setLinkStrategy(new CurvedLineLink<>());', caption="Minimal setup for <font face='Courier'>TreeNodeView</font>."),
            Figure(f"{G}/cover.svg", "A first look at the control."),
        ]),
        Chapter("Anatomy", [
            Para("The diagram and table identify the nodes, model objects and style classes that matter when using or styling the control."),
            Figure(f"{G}/anatomy.svg", "The main parts of the control."),
            Table(["Part", "Type / style", "Description"], [
                        ['TreeNodeView', 'tree-node-view', 'Control root and layout properties.'],
                        ['TreeNode', 'model node', 'Value, children, linkedNodes, expanded, width, height and name.'],
                        ['TreeNodeCell', 'tree-node-cell', 'Default cell with label and disclosure arrow.'],
                        ['tree-content', 'Group', 'Skin content group containing cells and link nodes.'],
                        ['LinkStrategy', 'link-line/path/curve/arrow', 'Draws parent-child and extra links.']
            ], widths=[20,30,50]),
        ]),
        Chapter("Control API", [
            Section('Tree and cells'),
            PropertyTable([
                        Property('root', 'ObjectProperty&lt;TreeNode&lt;T&gt;&gt;', 'null', 'Root node; null shows placeholder.'),
                        Property('cellFactory', 'ObjectProperty&lt;Callback&lt;T, TreeNodeCell&lt;T&gt;&gt;&gt;', 'TreeNodeCell::new', 'Creates cells for node values.'),
                        Property('placeholder', 'ObjectProperty&lt;Node&gt;', 'Label "No tree root."', 'Shown when root is null.'),
                        Property('refresh()', 'method', 'skin rebuild', 'Rebuilds visual tree from current model and properties.')
            ]),
            Section('Layout'),
            PropertyTable([
                        Property('cellWidth', 'DoubleProperty', '60', 'Default cell width; styleable.'),
                        Property('cellHeight', 'DoubleProperty', '30', 'Default cell height; styleable.'),
                        Property('hgap', 'DoubleProperty', '20', 'Horizontal gap; styleable.'),
                        Property('vgap', 'DoubleProperty', '50', 'Vertical gap; styleable.'),
                        Property('nodeLineGap', 'DoubleProperty', '10', 'Gap between cell and connecting lines; styleable.'),
                        Property('rowAlignment', 'ObjectProperty&lt;VPos&gt;', 'CENTER', 'Alignment for same-level nodes in vertical directions; styleable.'),
                        Property('columnAlignment', 'ObjectProperty&lt;HPos&gt;', 'CENTER', 'Alignment for same-level nodes in horizontal directions; styleable.'),
                        Property('layoutType', 'ObjectProperty&lt;LayoutType&gt;', 'REGULAR', 'REGULAR or COMPACT; styleable.'),
                        Property('layoutDirection', 'ObjectProperty&lt;LayoutDirection&gt;', 'TOP_TO_BOTTOM', 'Tree growth direction; styleable.')
            ]),
            Section('TreeNode'),
            PropertyTable([
                        Property('value', 'ObjectProperty&lt;T&gt;', 'null', 'User value.'),
                        Property('children', 'ObservableList&lt;TreeNode&lt;T&gt;&gt;', 'empty', 'Hierarchical child nodes; parent is maintained automatically.'),
                        Property('linkedNodes', 'ObservableList&lt;TreeNode&lt;T&gt;&gt;', 'empty', 'Additional non-hierarchical links.'),
                        Property('expanded', 'BooleanProperty', 'true', 'Controls visibility of descendants.'),
                        Property('width / height', 'DoubleProperty', 'USE_TREE_CELL_SIZE', 'Per-node dimensions; sentinel uses view defaults.'),
                        Property('name', 'String', 'null', 'Optional style identifier for node/link style classes.')
            ]),
        ]),
        Chapter("Behaviour", [
            Section('Layout algorithms'),
            Para('REGULAR layout sizes each subtree from child totals. COMPACT layout lays out levels breadth-first using the widest row or tallest column, which can reduce space.'),
            Figure(f"{G}/behaviour.svg", "The main runtime behaviour."),
            Section('Expansion and rebuilding'),
            Para('TreeNodeCell binds its expanded property bidirectionally to TreeNode.expanded. Changing expansion toggles descendant visibility and updates the tree.'),
            Figure(f"{G}/behaviour.svg", "Data and interaction flow."),
            Section('Links and style names'),
            Para('The skin asks linkStrategy to draw parent-child links and extra linkedNodes. If names exist, links receive classes such as link-parent-child or link-extra-source-target.'),
        ]),
        Chapter("Styling", [
            Para("The style hooks below were verified in the control, skin and CSS sources."),
            Figure(f"{G}/styling.svg", "Style hooks and visual states."),
            Section("Style classes"),
            Table(["Style class", "Where used"], [
                        ['tree-node-view', 'Root control style class.'],
                        ['tree-content', 'Group containing cells and links.'],
                        ['tree-node-cell', 'Default cell.'],
                        ['tree-node-cell-label', 'Label inside the default cell.'],
                        ['arrow-wrapper disclosure-arrow', 'Disclosure arrow shown when children exist.'],
                        ['link-arrow link-line link-path link-curve link-circle', 'Nodes returned by link strategies.'],
                        ['link-X-Y / link-extra-X-Y', 'Generated when TreeNode names are present.']
            ], widths=[35,65]),
            Section("Pseudo classes"),
            Table(["Pseudo class", "Meaning"], [
                        ['ltr rtl ttb btt', 'Direction pseudo classes on TreeNodeView.'],
                        ['expanded collapsed', 'TreeNodeCell expansion state.']
            ], widths=[30,70]),
            Section("Styleable CSS properties"),
            Table(["Property", "Type", "Default", "Description"], [
                        ['-fx-cell-width', 'Number', '60', 'Default cell width.'],
                        ['-fx-cell-height', 'Number', '30', 'Default cell height.'],
                        ['-fx-hgap', 'Number', '20', 'Horizontal gap.'],
                        ['-fx-vgap', 'Number', '50', 'Vertical gap.'],
                        ['-fx-node-line-gap', 'Number', '10', 'Gap between node and link.'],
                        ['-fx-row-alignment', 'VPos', 'CENTER', 'TOP, CENTER, BOTTOM or BASELINE.'],
                        ['-fx-column-alignment', 'HPos', 'CENTER', 'LEFT, CENTER or RIGHT.'],
                        ['-fx-layout-type', 'LayoutType', 'REGULAR', 'REGULAR or COMPACT.'],
                        ['-fx-layout-direction', 'LayoutDirection', 'TOP_TO_BOTTOM', 'LEFT_TO_RIGHT, RIGHT_TO_LEFT, TOP_TO_BOTTOM or BOTTOM_TO_TOP.']
            ], widths=[28,16,18,38]),
            Code('.tree-node-view {\n    -fx-layout-type: compact;\n    -fx-layout-direction: left-to-right;\n}\n.tree-node-view .link-line {\n    -fx-stroke: -fx-accent;\n}', caption="Example CSS."),
        ]),
        Chapter("Localization", [
            Para("The following keys are read via <font face='Courier'>ResourceBundleManager</font>."),
            Table(["Key", "English default"], [
                        ['placeholder.no-root', 'No tree root.']
            ], widths=[55,45]),
        ]),
        Chapter("Accessibility", [
            Para('TreeNodeView sets AccessibleRole.TREE_VIEW. The default TreeNodeCell does not set an explicit accessible role in its constructor.'),
        ]),
        Chapter("Recipes", [
            Figure(f"{G}/recipes.svg", "Common configuration recipes."),
            Section('Use compact left-to-right layout'),
            Code('view.setLayoutType(TreeNodeView.LayoutType.COMPACT);\nview.setLayoutDirection(TreeNodeView.LayoutDirection.LEFT_TO_RIGHT);'),
            Section('Per-node size'),
            Code('TreeNode<String> big = new TreeNode<>("Wide");\nbig.setSize(140, 50);'),
            Section('Add extra links'),
            Code('source.getLinkedNodes().add(target);\nsource.setName("source");\ntarget.setName("target");'),
            Section('Custom cell factory'),
            Code('view.setCellFactory(value -> {\n    TreeNodeCell<String> cell = new TreeNodeCell<>(value);\n    cell.getStyleClass().add("org-node");\n    return cell;\n});'),
            Section('Refresh after model-side changes'),
            Code('view.refresh();'),
            Section("Checklist"),
            Numbered(['Use TreeNode, not TreeItem.', 'Set TreeNode.name when you need stable link style classes.', 'Wrap the view in ScrollPane for large trees.', 'Choose rowAlignment only for vertical directions and columnAlignment for horizontal directions.']),
        ]),
        Chapter("See also", [
            Para("Demo app: <font face='Courier'>TreeNodeViewApp</font>. Run it with:"),
            Code("mvn javafx:run -f gemsfx-demo/pom.xml -Dmain.class=com.dlsc.gemsfx.demo.TreeNodeViewApp"),
            Bullets([
                'Related GemsFX controls: TreeNode, TreeNodeCell, link strategies, TreeView.',
                "API documentation: https://dlsc-software-consulting-gmbh.github.io/GemsFX/api/",
            ]),
        ]),
    ],
)
